ratatui-core/src/text/grapheme.rs

```diff
@@ -47,6 +55,99 @@ impl Styled for StyledGrapheme<'_> {
+}
+
+/// Calculates the width of a grapheme cluster as displayed in a terminal.
+///
+/// This function addresses a specific issue with halfwidth katakana dakuten and handakuten
+/// marks (U+FF9E ﾞ and U+FF9F ﾟ). While `unicode-width` reports these as width 0 (due to
+/// their `Grapheme_Extend` property), terminals display them as independent characters with
+/// width 1 each, inherited from legacy Shift-JIS encoding where they were separate characters.
+///
+/// # Background
+///
+/// In legacy Shift-JIS encoding (JIS X 0201), halfwidth katakana and dakuten/handakuten
+/// were completely separate 1-byte characters. This behavior persists in terminal emulators
+/// for compatibility, where `ｶﾞ` is rendered as two distinct character cells, not as a
+/// combined single character.
+///
+/// The combining variants (U+3099 and U+309A) properly combine with preceding characters
+/// and report correct width 0, so they don't need special handling.
+fn terminal_width(grapheme: &str) -> usize {
+    // Check only the last character because halfwidth dakuten/handakuten (U+FF9E/U+FF9F)
+    // always appear at the END of a grapheme cluster, never at the beginning or middle.
+    // This is guaranteed by Unicode normalization rules:
+    // - Characters with Grapheme_Extend property (including these marks) must follow their base
+    //   character
+    // - If a dakuten appears at the start, it forms a separate grapheme cluster
+    // Example: "ﾞｶ" becomes two clusters: "ﾞ" (alone) + "ｶ" (alone), not "ﾞｶ" (combined)
+    if let Some(c) = grapheme.chars().last() {
+        if is_halfwidth_dakuten_handakuten(c) {
+            // Sum up the width of each character individually
+            return grapheme
+                .chars()
+                .map(|c| {
+                    if is_halfwidth_dakuten_handakuten(c) {
+                        // Halfwidth dakuten/handakuten are width 1 in terminals
+                        1
+                    } else {
+                        // Use unicode-width for other characters
+                        // unwrap_or(1) handles control characters and other edge cases
+                        c.width().unwrap_or(1)
+                    }
+                })
+                .sum();
+        }
+    }
+
+    // For all other cases, use unicode-width as-is
+    grapheme.width()
+}
+
+/// Extension trait to calculate terminal display width for strings.
+///
+/// This trait provides a `.terminal_width()` method for `&str` that correctly handles
+/// halfwidth katakana dakuten and handakuten marks, similar to how `UnicodeWidthStr`
+/// provides `.width()`.
+///
+/// # Examples
+///
+/// ```
+/// use ratatui_core::text::grapheme::TerminalWidthStr;
+///
+/// assert_eq!("hello".terminal_width(), 5);
+/// assert_eq!("ｶﾞｷﾞｸﾞ".terminal_width(), 6); // 3 graphemes × 2 width each
+/// assert_eq!("あいう".terminal_width(), 6); // 3 graphemes × 2 width each
+/// ```
+pub trait TerminalWidthStr {
```

Comment:
Small nit: I think this and the above function belongs in a new file that's just about the width. Name it `width.rs`
That keeps the testing small and focused around this one aspect of things. There's not a lot of overlap on the grapheme code here.